dev-tools/releng/src/job.rs

```diff
@@ -0,0 +1,284 @@
+            ::slog::error!($logger, $($tt)*);
+        }
+    };
+}
+
+async fn run_job<F>(
+    logger: Logger,
+    permits: Arc<Semaphore>,
+    name: String,
+    future: F,
+) -> Result<()>
+where
+    F: Future<Output = Result<()>> + 'static,
+{
+    if !PERMIT_NOT_REQUIRED.contains(&name.as_str()) {
+        let _ = permits.acquire_owned().await?;
+    }
+
+    info!(logger, "[{}] running task", name);
+    let start = Instant::now();
+    let result = future.await;
+    let duration = Instant::now().saturating_duration_since(start);
+    info_or_error!(
+        logger,
+        result,
+        "[{}] task {} ({:?})",
+        name,
+        if result.is_ok() { "succeeded" } else { "failed" },
+        duration
+    );
+    result
+}
+
+async fn spawn_with_output(
+    mut command: Command,
+    logger: Logger,
+    permits: Arc<Semaphore>,
+    name: String,
+    log_path: Utf8PathBuf,
+) -> Result<()> {
+    let _ = permits.acquire_owned().await?;
+
+    let log_file_1 = File::create(log_path).await?;
+    let log_file_2 = log_file_1.try_clone().await?;
+
+    info!(logger, "[{}] running: {}", name, command.to_string());
+    let start = Instant::now();
+    let mut child = command
+        .kill_on_drop(true)
+        .stdin(Stdio::null())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()
+        .with_context(|| format!("failed to exec `{}`", command.to_string()))?;
+
+    let stdout = reader(
+        &name,
+        child.stdout.take().unwrap(),
+        tokio::io::stdout(),
+        log_file_1,
+    );
+    let stderr = reader(
+        &name,
+        child.stderr.take().unwrap(),
+        tokio::io::stderr(),
+        log_file_2,
+    );
+    match tokio::try_join!(child.wait(), stdout, stderr) {
+        Ok((status, (), ())) => {
+            let result = command.check_status(status);
+            info_or_error!(
+                logger,
+                result,
+                "[{}] process exited with {} ({:?})",
+                name,
+                status,
+                Instant::now().saturating_duration_since(start)
+            );
+            result
+        }
+        Err(err) => Err(err).with_context(|| {
+            format!("I/O error while waiting for job {:?} to complete", name)
+        }),
+    }
+}
+
+async fn reader(
+    name: &str,
+    reader: impl AsyncRead + Unpin,
+    mut terminal_writer: impl AsyncWrite + Unpin,
+    logfile_writer: File,
+) -> std::io::Result<()> {
+    let mut reader = BufReader::new(reader);
+    let mut logfile_writer = tokio::fs::File::from(logfile_writer);
+    let mut buf = format!("[{:>16}] ", name).into_bytes();
+    let prefix_len = buf.len();
+    loop {
+        buf.truncate(prefix_len);
+        let size = reader.read_until(b'\n', &mut buf).await?;
+        if size == 0 {
+            return Ok(());
+        }
+        terminal_writer.write_all(&buf).await?;
+        logfile_writer.write_all(&buf[prefix_len..]).await?;
+    }
```

Comment:
Is this the same as https://docs.rs/tokio-stream/0.1.15/tokio_stream/wrappers/struct.LinesStream.html?

Reply:
Aha, good catch. Agreed that we shouldn't outright fail on non-UTF-8 output. (Again, can this be put in a comment?)